src/app.js

```diff
@@ -247,4 +229,114 @@ export default function (app) {
 
     return element
   }
+
+  function patchChildren(element, oldNode, node) {
+    var len = node.children.length
+    var oldLen = oldNode.children.length
+    var reusableChildren = {}
+    var oldElements = []
+    var newKeys = {}
+    var oldKey
+    var i = 0
+    var j = 0
+
+    for (var i = 0; i < oldLen; i++) {
+      var oldElement = element.childNodes[i]
+      oldElements[i] = oldElement
+
+      var oldChild = oldNode.children[i]
+
+      if (oldChild.data && (oldKey = oldChild.data.key)) {
+        reusableChildren[oldKey] = [oldElement, oldChild]
+      }
+    }
+
+    while (j < len) {
+      var oldElement = oldElements[i]
+      var oldChild = oldNode.children[i]
+      var newChild = node.children[j]
+
+      // debugger
+
+      var oldKey = 0
+      var newKey = 0
+
+      if (oldChild && oldChild.data) {
+        oldKey = oldChild.data.key
+      }
+
+      if (newChild.data) {
+        newKey = newChild.data.key
+      }
+
+      var reusableChild = reusableChildren[newKey]
+
+      var reusableElement = 0
+      var reusableNode = 0
+
+      if (reusableChild) {
+        reusableElement = reusableChild[0]
+        reusableNode = reusableChild[1]
+      }
+
+      if (!oldKey && !newKey) {
+        patch(element, oldElement, oldChild, newChild)
+        j++
+        i++
+
+      } else if (!oldKey && newKey) {
+        if (reusableElement) {
+          element.insertBefore(reusableElement, oldElement)
+          patch(element, reusableElement, reusableNode, newChild)
+        } else {
+          patch(element, oldElement, null, newChild)
+        }
+        j++
+        newKeys[newKey] = true
+
+      } else if (oldKey && !newKey) {
```

Comment:
just checking `!newKey` might result in false positives for keys such as `0` or `''` (not sure we should allow empty string as a key. But perhaps 0 should be ok.